ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/AddMetadata.java

```diff
@@ -4,45 +4,173 @@
+    Optional.ofNullable(metadata).map(m -> m.get("user_agent")) //
+        .filter(Map.class::isInstance) //
+        .map(m -> ((Map<String, Object>) m).entrySet()) //
+        .orElse(ImmutableSet.of()) //
+        .forEach(
+            entry -> attributes.put("user_agent_" + entry.getKey(), entry.getValue().toString()));
+  }
+
+  static Map<String, Object> headersFromAttributes(Map<String, String> attributes) {
+    HashMap<String, Object> header = new HashMap<>();
+    HEADER_ATTRIBUTES.stream().forEach(name -> Optional //
+        .ofNullable(attributes.get(name)) //
+        .ifPresent(value -> header.put(name, value)));
+    return header;
+  }
+
+  static void putHeaderAttributes(Map<String, String> attributes, Map<String, Object> metadata) {
+    Optional.ofNullable(metadata).map(m -> m.get("header")) //
+        .filter(Map.class::isInstance) //
+        .map(m -> ((Map<String, Object>) m).entrySet()) //
+        .orElse(ImmutableSet.of()) //
+        .forEach(entry -> attributes.put(entry.getKey(), entry.getValue().toString()));
+  }
+
+  static Map<String, Object> uriFromAttributes(Map<String, String> attributes) {
+    HashMap<String, Object> uri = new HashMap<>();
+    URI_ATTRIBUTES.stream().forEach(name -> Optional //
+        .ofNullable(attributes.get(name)) //
+        .ifPresent(value -> uri.put(name, value)));
+    return uri;
+  }
+
+  static void putUriAttributes(Map<String, String> attributes, Map<String, Object> metadata) {
+    Optional.ofNullable(metadata).map(m -> m.get("uri")) //
+        .filter(Map.class::isInstance) //
+        .map(m -> ((Map<String, Object>) m).entrySet()) //
+        .orElse(ImmutableSet.of()) //
+        .forEach(entry -> attributes.put(entry.getKey(), entry.getValue().toString()));
+  }
+
   @Override
   protected PubsubMessage processElement(PubsubMessage message) throws IOException {
     message = PubsubConstraints.ensureNonNull(message);
     // Get payload
     final byte[] payload = message.getPayload();
     // Get attributes as bytes, throws IOException
-    final byte[] metadata = Json.asBytes(message.getAttributeMap());
+    final byte[] metadata = Json.asBytes(attributesToMetadataPayload(message.getAttributeMap()));
     // Ensure that we have a json object with no leading whitespace
     if (payload.length < 2 || payload[0] != '{') {
       throw new IOException("invalid json object: must start with {");
     }
     // Create an output stream for joining metadata with payload
     final ByteArrayOutputStream payloadWithMetadata = new ByteArrayOutputStream(
-        METADATA_PREFIX.length + metadata.length + payload.length);
-    // Write metadata prefix
-    payloadWithMetadata.write(METADATA_PREFIX);
-    // Write metadata
-    payloadWithMetadata.write(metadata);
+        metadata.length + payload.length);
+    // Write metadata without trailing `}`
+    payloadWithMetadata.write(metadata, 0, metadata.length - 1);
```

Comment:
probably out of scope, maybe completely unnecessary:

someone mentioned conflicting schemas overriding metadata values with payload values because of repeated keys (e.g. `{"metadata":4}` -> `{"metadata":{...},"metdata":4}`), and I don't know why it didn't occur to me before, but we could give attributes the highest priority by appending them instead of prepending them (e.g. `{"metadata":4}` -> `{"metdata":4,"metadata":{...}}`)

```diff
-    // Write metadata without trailing `}`
-    payloadWithMetadata.write(metadata, 0, metadata.length - 1);
+    // Write payload without trailing `}`
+    payloadWithMetadata.write(payload, 0, payload.length - 1);
    // Start next json field, unless object was empty
    if (payload.length > 2) {
      // Write comma to start the next field
      payloadWithMetadata.write(',');
    }
-    // Write payload without leading `{`
-    payloadWithMetadata.write(payload, 1, payload.length - 1);
+    // Write metadata without leading `{`
+    payloadWithMetadata.write(metadata, 1, metadata.length - 1);
```

Reply:
Either way, we end up with invalid json due to the duplicated keys, yes?

Given the assumption that AddMetadata will only ever happen in a pipeline after ParsePayload has been applied, we should never run into this situation. ParsePayload calls `stripPayloadMetadataToAttributes` which will overwrite any existing attributes with reserved names and remove all fields from the payload that are named like metadata fields. Nothing else alters the payload before hitting AddMetadata, so we should be guaranteed that none of the fields we're adding back in are already present in the payload.

But, having said this, I realize that I did not actually add the call to `stripPayloadMetadataToAttributes` into ParsePayload, so that's a major bug I will fix forthwith.